THREESCALE-10388 Migrate apicast DeploymentConfigs to Deployments

carlkyrillos opened the pull request:
# Issue link
JIRA: [THREESCALE-10388](https://issues.redhat.com/browse/THREESCALE-10388)

# What
This is the first PR against the new feature branch `threescale-10163` that will be used while implementing the changes needed to migrate each of 3scale-operator's DeploymentConfigs to Deployments - see [epic](https://issues.redhat.com/browse/THREESCALE-10163).

This PR replaces the `apicast-staging` and `apicast-production` DeploymentConfigs with Deployments. The PR also provides the upgrade scaffolding that will be used by the remaining DeploymentConfigs.

# Verification steps

1. Provision a cluster
2. Install 3scale locally using master
3. Checkout this PR
4. Re-run 3scale locally
5. Verify that the new `apicast-staging` and `apicast-production` Deployments are created
6. Verify that the `apicast-staging` and `apicast-production` DeploymentConfigs are deleted once the Deployments report healthy
7. Login to admin portal and create dummy product/backend.
8. Promote the APIcast configuration to Staging APIcast and curl the staging endpoint with a valid user_key to confirm it's working
9. Promote the APIcast configuration to Production APIcast and curl the production endpoint with a valid user_key to confirm it's working

austincunningham commented:
# Verification
- [x] Provision a cluster
- [x] Install 3scale locally using master
- [x] Checkout this PR
- [x] Re-run 3scale locally
seeing this error 
```bash
1.7005637337319179e+09  ERROR   Reconciler error        {"controller": "apimanager", "controllerGroup": "apps.3scale.net", "controllerKind": "APIManager", "aPIManager": {"name":"apimanager-sample","namespace":"3scale-test"}, "namespace": "3scale-test", "name": "apimanager-sample", "reconcileID": "543cd136-1419-4019-88be-43487fe9722d", "error": "deployment apicast-staging is not yet available"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        /home/austincunningham/.gvm/pkgsets/go1.19/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        /home/austincunningham/.gvm/pkgsets/go1.19/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:234
```
although it reconciles out eventually 
- [x] Verify that the new apicast-staging and apicast-production Deployments are created
```bash
oc get deploy | grep apicast
apicast-production   1/1     1            1           5m39s
apicast-staging      1/1     1            1           6m10s
# get the status
oc get deploy apicast-production -oyaml | yq '.status'
```
```yaml
availableReplicas: 1
conditions:
  - lastTransitionTime: "2023-11-21T10:49:42Z"
    lastUpdateTime: "2023-11-21T10:49:42Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2023-11-21T10:49:11Z"
    lastUpdateTime: "2023-11-21T10:49:42Z"
    message: ReplicaSet "apicast-production-58c5cfdf99" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
observedGeneration: 2
readyReplicas: 1
replicas: 1
updatedReplicas: 1
```
```bash
oc get deploy apicast-staging -oyaml | yq '.status'
```
```yaml
availableReplicas: 1
conditions:
  - lastTransitionTime: "2023-11-21T10:49:11Z"
    lastUpdateTime: "2023-11-21T10:49:11Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2023-11-21T10:48:40Z"
    lastUpdateTime: "2023-11-21T10:49:11Z"
    message: ReplicaSet "apicast-staging-7996f498cf" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
observedGeneration: 2
readyReplicas: 1
replicas: 1
updatedReplicas: 1

```
- [x] Verify that the apicast-staging and apicast-production DeploymentConfigs are deleted once the Deployments report healthy
```bash
oc get dc | grep apicast
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+

```
- [x] Login to admin portal and create dummy product/backend.
 used the wizard
- [x] Promote the APIcast configuration to Staging APIcast and curl the staging endpoint with a valid user_key to confirm it's working
```bash
curl "https://api-3scale-apicast-staging.apps.aucunnin.lpi0.s1.devshift.org:443/?user_key=3500ae084aaa902816d9f3b3b37dba23"
{
  "method": "GET",
  "path": "/",
  "args": "user_key=3500ae084aaa902816d9f3b3b37dba23",
  "body": "",
  "headers": {
    "HTTP_VERSION": "HTTP/1.1",
    "HTTP_HOST": "echo-api.3scale.net",
    "HTTP_X_REAL_IP": "10.131.2.7",
    "HTTP_X_3SCALE_PROXY_SECRET_TOKEN": "Shared_secret_sent_from_proxy_to_API_backend_50bd6dc03615844e",
    "HTTP_USER_AGENT": "curl/7.85.0",
    "HTTP_ACCEPT": "*/*",
    "HTTP_X_FORWARDED_HOST": "api-3scale-apicast-staging.apps.aucunnin.lpi0.s1.devshift.org",
    "HTTP_X_FORWARDED_PORT": "443",
    "HTTP_X_FORWARDED_PROTO": "https",
    "HTTP_FORWARDED": "for=185.167.180.230;host=api-3scale-apicast-staging.apps.aucunnin.lpi0.s1.devshift.org;proto=https",
    "HTTP_X_FORWARDED_FOR": "185.167.180.230,44.218.61.13",
    "HTTP_X_ENVOY_EXTERNAL_ADDRESS": "44.218.61.13",
    "HTTP_X_REQUEST_ID": "2ac93670-7e58-424c-a8fc-bbfef14c8618",
    "HTTP_X_ENVOY_EXPECTED_RQ_TIMEOUT_MS": "15000"
  },
  "uuid": "f4e58cfd-e23d-478b-9a24-f83c3d1d4eae"
}%       
```
- [x] Promote the APIcast configuration to Production APIcast and curl the production endpoint with a valid user_key to confirm it's working
Production failed :thinking: 
actually just took a while :shrug: 
```bash
curl "https://api-3scale-apicast-production.apps.aucunnin.lpi0.s1.devshift.org:443/?user_key=3500ae084aaa902816d9f3b3b37dba23"
{
  "method": "GET",
  "path": "/",
  "args": "user_key=3500ae084aaa902816d9f3b3b37dba23",
  "body": "",
  "headers": {
    "HTTP_VERSION": "HTTP/1.1",
    "HTTP_HOST": "echo-api.3scale.net",
    "HTTP_X_REAL_IP": "10.128.4.10",
    "HTTP_X_3SCALE_PROXY_SECRET_TOKEN": "Shared_secret_sent_from_proxy_to_API_backend_50bd6dc03615844e",
    "HTTP_USER_AGENT": "curl/7.85.0",
    "HTTP_ACCEPT": "*/*",
    "HTTP_X_FORWARDED_HOST": "api-3scale-apicast-production.apps.aucunnin.lpi0.s1.devshift.org",
    "HTTP_X_FORWARDED_PORT": "443",
    "HTTP_X_FORWARDED_PROTO": "https",
    "HTTP_FORWARDED": "for=185.167.180.230;host=api-3scale-apicast-production.apps.aucunnin.lpi0.s1.devshift.org;proto=https",
    "HTTP_X_FORWARDED_FOR": "185.167.180.230,44.218.61.13",
    "HTTP_X_ENVOY_EXTERNAL_ADDRESS": "44.218.61.13",
    "HTTP_X_REQUEST_ID": "480e5cac-aee6-4788-8237-c5c704b4636c",
    "HTTP_X_ENVOY_EXPECTED_RQ_TIMEOUT_MS": "15000"
  },
  "uuid": "54b6fa01-a9e0-4a22-a0cf-3a514e4cc2f3"
}%                                                  
```

austincunningham commented:
Might need a rebase against master to pull in the TOC fix

austincunningham commented:
> @austincunningham Thanks for the review, a few quick things:
> 
> * That reconciler error about the Deployment not being ready is intended but I can change the log level to DEBUG.
Sounds good  :heavy_check_mark: 
> * How long did it take for the apicast-production test to work? There shouldn't be any downtime because the DC shouldn't be deleted until the Deployment is fully healthy.
It took about a minute or so,  Not caused by the upgrade only tried the promotion after we were using deployments. 
> * Will rebase to pull in the TOC fix
:heavy_check_mark:

eguzki commented:
recently I learned that a github generates a TOC automatically for all README.md. So we can remove all the generation tooling and validations regarding the TOC.

carlkyrillos commented:
@austincunningham I rebased off of master and updated the logging level. Can you take another look at this when you have a chance?

eguzki commented on pkg/3scale/amp/component/apicast.go:
```diff
@@ -725,17 +695,18 @@ func (apicast *Apicast) productionDeploymentConfigAnnotations() map[string]strin
+	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString(map[string]string{
+		ApicastStagingName: fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+	})
+	annotations[reconcilers.DeploymentImageTriggerAnnotation] = imageTriggerString
```
What is the purpose of this new annotation?

carlkyrillos replied:
This is needed for the Deployment objects to leverage the existing ImageStreams. See docs [here](https://docs.openshift.com/container-platform/4.14/openshift_images/triggering-updates-on-imagestream-changes.html) for further explanation.

eguzki replied:
today I learned this from Openshift. Thanks

Have you considered not using triggers? What is the benefit you see in using triggers?

carlkyrillos replied:
My understanding is that we have to use triggers in order to use the existing ImageStreams. I'm fine with skipping triggers/ImageStreams but would we then need the operator to handle image versioning and patching the Deployments with each version upgrade?

eguzki replied:
the operator is needed to handle image versioning and patching the deployments regardless of using imagestreams or not. 

The question is, do we want to keep using Imagestreams? and if so, what is the benefit?

carlkyrillos replied:
I don't have a strong opinion. ImageStreams contain some metadata, image history, etc. but not sure if that's good enough of a reason to keep them though.

Do you know what the original justification for using ImageStreams was? Was it just because DCs required them for triggers?

carlkyrillos replied:
@eguzki Would there ever be a scenario where a component's image would be updated outside of a 3scale-operator version bump? If so, then I think we need to continue to use ImageStreams to sync the updated component image to the Deployment, right?

eguzki replied:
I do not think such scenario is realistic. It will not be supported use case. And for that, there are `Image` fields to update images using the APIMAnager CR.

I tend to think we can get rid of the imagestreams. It can be done in following up PRs

carlkyrillos replied:
Sounds good I'll merge this and will address it in another PR.

carlkyrillos commented:
@eguzki Thank you for the heads up about the TOC validation. 

Does the apicast DC migration logic look good to you? If so, can we get this merged to the feature branch? I have branches ready for backend-listener|cron|worker and system-memcache (although I will first need to update them as the latest changes to master changed the DeploymentConfig mutators and backend reconciliation logic)

eguzki commented on pkg/3scale/amp/operator/backend_reconciler.go:
```diff
@@ -38,7 +41,19 @@ func (r *BackendReconciler) Reconcile() (reconcile.Result, error) {
 
 	// Listener DC
+	backendRedisSecret := &v1.Secret{}
+	r.Client().Get(r.Context(), types.NamespacedName{
```
Are these changes related to the current PR?

Additionally, what if the client returns error? error handling is missing?

carlkyrillos replied:
No these changes are from #862. I rebased off of master to pull in the TOC validation fix.

eguzki commented on pkg/reconcilers/deploymentconfig.go:
```diff
@@ -57,6 +57,84 @@ func GenericBackendMutators() []DCMutateFn {
 	}
 }
 
+func DeploymentConfigListenerArgsMutator(_, existing *appsv1.DeploymentConfig) (bool, error) {
```
I guess some unrelated content has been sneaked in.

carlkyrillos replied:
Yes this is from #862. I can rebase the [base feature branch](https://github.com/3scale/3scale-operator/tree/threescale-10163) off of master instead in order to make this PR's changes clearer. @eguzki just let me know which you'd prefer

eguzki commented on pkg/upgrade/migrate_deployment_config_2.15.go:
```diff
@@ -0,0 +1,55 @@
+	if k8serr.IsNotFound(err) {
+		return true, nil
+	}
+	if err != nil && !k8serr.IsNotFound(err) {
```
nitpicking here
```go
if err != nil  {
    ...
}
```
should be enough, right?

carlkyrillos replied:
Good catch, will change

eguzki commented on pkg/upgrade/migrate_deployment_config_2.15.go:
```diff
@@ -0,0 +1,55 @@
+
+	// Delete the DeploymentConfig because the Deployment replacing it is healthy
+	err = client.Delete(context.TODO(), deploymentConfig)
+	if err != nil {
```
very unlikely, but it may return not found error. I would

```go
if err != nil {
    if !k8serr.IsNotFound(err) {
          return false, fmt.Errorf("error deleting deploymentconfig %s: %v", deploymentConfig.Name, err)
    }
}
```

carlkyrillos replied:
Good catch, will change

eguzki reviewed (changes requested):
Looking good. 

I give my :+1: to the approach followed for the migration.

eguzki commented on pkg/3scale/amp/component/apicast.go:
```diff
@@ -142,7 +125,7 @@ func (apicast *Apicast) StagingDeploymentConfig() *appsv1.DeploymentConfig {
+						{
 							Ports:           apicast.stagingContainerPorts(),
 							Env:             apicast.buildApicastStagingEnv(),
 							Image:           "amp-apicast:latest",
```
Should we set the image?

carlkyrillos replied:
I kept it because that's how the DeploymentConfig had it configured but I don't think we have to set it. Either way it will be changed due to the ImageStreamTag annotation - see [docs](https://docs.openshift.com/container-platform/4.14/openshift_images/triggering-updates-on-imagestream-changes.html).

carlkyrillos commented:
@austincunningham @eguzki I incorporated the latest suggestions and cleaned up the commits. 

If there are no objections I think this can be merged and I'll get the next few PRs to the feature branch ready.

eguzki commented on pkg/reconcilers/deployment.go:
```diff
@@ -0,0 +1,272 @@
+
+	annotationString := ""
+	if len(containers) > 0 {
+		annotationString += `[`
```
Even though it is correct, what about this alternative implementation (feel free to customize it as you wish)

```diff
diff --git a/pkg/3scale/amp/component/apicast.go b/pkg/3scale/amp/component/apicast.go
index 9b8a102c..9dcd90fd 100644
--- a/pkg/3scale/amp/component/apicast.go
+++ b/pkg/3scale/amp/component/apicast.go
@@ -676,9 +676,15 @@ func (apicast *Apicast) productionDeploymentAnnotations() map[string]string {
 	annotations := map[string]string{}
 
 	// Image trigger annotation should always be present
-	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString(map[string]string{
-		ApicastProductionInitContainerName: fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
-		ApicastProductionName:              fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString([]reconcilers.ContainerImage{
+		{
+			Name: ApicastProductionInitContainerName,
+			Tag:  fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+		},
+		{
+			Name: ApicastProductionName,
+			Tag:  fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+		},
 	})
 	annotations[reconcilers.DeploymentImageTriggerAnnotation] = imageTriggerString
 
@@ -702,8 +708,11 @@ func (apicast *Apicast) stagingDeploymentAnnotations() map[string]string {
 	annotations := map[string]string{}
 
 	// Image trigger annotation should always be present
-	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString(map[string]string{
-		ApicastStagingName: fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+	imageTriggerString := reconcilers.CreateImageTriggerAnnotationString([]reconcilers.ContainerImage{
+		{
+			Name: ApicastStagingName,
+			Tag:  fmt.Sprintf("amp-apicast:%v", apicast.Options.ImageTag),
+		},
 	})
 	annotations[reconcilers.DeploymentImageTriggerAnnotation] = imageTriggerString
 
diff --git a/pkg/reconcilers/deployment.go b/pkg/reconcilers/deployment.go
index eb3f459d..b8dc3ccc 100644
--- a/pkg/reconcilers/deployment.go
+++ b/pkg/reconcilers/deployment.go
@@ -1,10 +1,9 @@
 package reconcilers
 
 import (
+	"encoding/json"
 	"fmt"
 	"reflect"
-	"sort"
-	"strings"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
@@ -249,24 +248,44 @@ func DeploymentPodTemplateAnnotationsMutator(desired, existing *k8sappsv1.Deploy
 	return updated, nil
 }
 
+type ContainerImage struct {
+	Name string
+	Tag  string
+}
+
+type ImageTriggerFrom struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+	// +optional
+	Namespace *string `json:"namespace,omitempty"`
+}
+
+type ImageTrigger struct {
+	From      ImageTriggerFrom `json:"from"`
+	FieldPath string           `json:"fieldPath"`
+	// +optional
+	Pused bool `json:"paused,omitempty"`
+}
+
 // CreateImageTriggerAnnotationString creates the annotation for Deployments to leverage existing ImageStreamTags
-func CreateImageTriggerAnnotationString(containers map[string]string) string {
-	// Need to sort the map to guarantee the range iteration order, otherwise the operator may unnecessarily update the annotations.
-	var keys []string
-	for k := range containers {
-		keys = append(keys, k)
+func CreateImageTriggerAnnotationString(containers []ContainerImage) string {
+
+	triggers := make([]ImageTrigger, len(containers))
+
+	for _, container := range containers {
+		triggers = append(triggers, ImageTrigger{
+			From: ImageTriggerFrom{
+				Kind: "ImageStreamTag",
+				Name: container.Tag,
+			},
+			FieldPath: fmt.Sprintf(`spec.template.spec.containers[?(@.name=="%v")].image`, container.Name),
+		})
 	}
-	sort.Strings(keys)
 
-	annotationString := ""
-	if len(containers) > 0 {
-		annotationString += `[`
-		for _, c := range keys {
-			annotationString += fmt.Sprintf(`{"from":{"kind":"ImageStreamTag","name":"%v"},"fieldPath":"spec.template.spec.containers[?(@.name==\"%v\")].image"},`, containers[c], c)
-		}
-		annotationString = strings.TrimSuffix(annotationString, ",")
-		annotationString += `]`
+	if len(triggers) > 0 {
+		jsonData, _ := json.Marshal(triggers)
+		return string(jsonData)
 	}
 
-	return annotationString
+	return ""
 }
```

carlkyrillos replied:
Thank you for this, this is a much more elegant solution. I just tested and pushed the changes.

austincunningham commented:
> recently I learned that a github generates a TOC automatically for all README.md. So we can remove all the generation tooling and validations regarding the TOC.
> 
> (In another PR, obviously)

Yep once we have 2 or more headers they will be auto generated by github, Who knew :scream: 
https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax
